miden-crypto/src/encryption/aead_rpo/mod.rs

```diff
@@ -0,0 +1,477 @@
+
+/// Second half of the rate portion of the sponge state
+const RATE_RANGE_SECOND_HALF: Range<usize> =
+    Rpo256::RATE_RANGE.start + HALF_RATE_WIDTH..Rpo256::RATE_RANGE.end;
+
+/// Index of the first element of the rate portion of the sponge state
+const RATE_START: usize = Rpo256::RATE_RANGE.start;
+
+/// Padding block used when the length of the data to encrypt is a multiple of `RATE_WIDTH`
+const PADDING_BLOCK: [Felt; RATE_WIDTH] = [ONE, ZERO, ZERO, ZERO, ZERO, ZERO, ZERO, ZERO];
+
+/// Number of bytes to pack into one field element
+const BINARY_CHUNK_SIZE: usize = 7;
+
+// TYPES AND STRUCTURES
+// ================================================================================================
+
+/// Encrypted data with its authentication tag
+#[derive(Debug, PartialEq, Eq)]
+pub struct EncryptedData {
+    associated_data: Vec<Felt>,
+    ciphertext: Vec<Felt>,
+    auth_tag: AuthTag,
+    nonce: Nonce,
+}
+
+impl EncryptedData {
+    pub fn associated_data(&self) -> &[Felt] {
+        &self.associated_data
+    }
+}
+
+/// An authentication tag represented as 4 field elements
+#[derive(Debug, Default, Clone, PartialEq, Eq)]
+pub struct AuthTag([Felt; AUTH_TAG_SIZE]);
+
+/// A 256-bit secret key represented as 4 field elements
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub struct SecretKey([Felt; SECRET_KEY_SIZE]);
+
+impl SecretKey {
+    /// Creates a new random secret key using the default random number generator
+    #[cfg(feature = "std")]
+    #[allow(clippy::new_without_default)]
+    pub fn new() -> Self {
+        use rand::{SeedableRng, rngs::StdRng};
+        let mut rng = StdRng::from_os_rng();
+
+        Self::with_rng(&mut rng)
+    }
+
+    /// Creates a new random secret key using the provided random number generator
+    pub fn with_rng<R: Rng>(rng: &mut R) -> Self {
+        rng.sample(StandardUniform)
+    }
+
+    /// Encrypts the provided data, as Felt-s, using this secret key and a random nonce
+    #[cfg(feature = "std")]
+    pub fn encrypt(&self, data: &[Felt], associated_data: &[Felt]) -> EncryptedData {
+        use rand::{SeedableRng, rngs::StdRng};
+        let mut rng = StdRng::from_os_rng();
+        let nonce = Nonce::with_rng(&mut rng);
+
+        self.encrypt_with_nonce(data, associated_data, nonce)
+    }
+
+    /// Encrypts the provided data, as bytes, using this secret key and a random nonce
+    #[cfg(feature = "std")]
+    pub fn encrypt_bytes(&self, data: &[u8], associated_data: &[u8]) -> EncryptedData {
+        use rand::{SeedableRng, rngs::StdRng};
+        let mut rng = StdRng::from_os_rng();
+        let nonce = Nonce::with_rng(&mut rng);
+
+        self.encrypt_bytes_with_nonce(data, associated_data, nonce)
+    }
+
+    /// Encrypts the provided data using this secret key and a specified nonce
+    pub fn encrypt_with_nonce(
+        &self,
+        data: &[Felt],
+        associated_data: &[Felt],
+        nonce: Nonce,
+    ) -> EncryptedData {
```

Comment:
Would be great to explain in the doc comments how padding works both for data and associated data. I think the main question here is whether we get different cyphertext for things like `[ONE]`, and `[ONE, ZERO]` (both for data and associated data).

Reply:
Done